File: validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def recursiveValidBST(self, node: Optional[TreeNode], lower: int, upper: int) ->  bool:
        if not node:
            return True
        print(node.val, lower, upper)
        if node.val <= lower or node.val >= upper:
            return False
        return self.recursiveValidBST(node.left, lower, node.val) and self.recursiveValidBST(node.right, node.val, upper)


    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        return self.recursiveValidBST(root.left, float('-inf'), root.val) and self.recursiveValidBST(root.right, root.val, float('inf'))
  

File: c++/validate-binary-search-tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if (root == nullptr) return true;
        stack<TreeNode*> nodeStack;
        TreeNode* prev = nullptr;
        while (root != nullptr || !nodeStack.empty()) {
            while (root != nullptr) {
                nodeStack.push(root);
                root = root->left;
            }
            root = nodeStack.top();
            nodeStack.pop();
            if (prev != nullptr && root->val <= prev->val) return false;
            prev = root;
            root = root->right;
        }
        return true;
    }
};